crates/chia-datalayer/src/merkle/blob.rs

```diff
@@ -0,0 +1,2557 @@
+                },
+                node: Node::Leaf(node),
+            };
+
+            self.insert_entry_to_blob(index, &block)?;
+        }
+
+        Ok(nodes[1].0)
+    }
+
+    fn insert_third_or_later(
+        &mut self,
+        mut node: LeafNode,
+        old_leaf: &LeafNode,
+        old_leaf_index: TreeIndex,
+        internal_node_hash: &Hash,
+        side: Side,
+    ) -> Result<TreeIndex, Error> {
+        let new_leaf_index = self.get_new_index();
+        let new_internal_node_index = self.get_new_index();
+
+        node.parent = Parent(Some(new_internal_node_index));
+
+        let new_leaf_block = Block {
+            metadata: NodeMetadata {
+                node_type: NodeType::Leaf,
+                dirty: false,
+            },
+            node: Node::Leaf(node),
+        };
+        self.insert_entry_to_blob(new_leaf_index, &new_leaf_block)?;
+
+        let (left_index, right_index) = match side {
+            Side::Left => (new_leaf_index, old_leaf_index),
+            Side::Right => (old_leaf_index, new_leaf_index),
+        };
+        let new_internal_block = Block {
+            metadata: NodeMetadata {
+                node_type: NodeType::Internal,
+                dirty: false,
+            },
+            node: Node::Internal(InternalNode {
+                parent: old_leaf.parent,
+                left: left_index,
+                right: right_index,
+                hash: *internal_node_hash,
+            }),
+        };
+        self.insert_entry_to_blob(new_internal_node_index, &new_internal_block)?;
+
+        let old_parent_index = old_leaf.parent.0.expect("root found when not expected");
+
+        self.update_parent(old_leaf_index, Some(new_internal_node_index))?;
+
+        let mut old_parent_block = self.get_block(old_parent_index)?;
+        if let Node::Internal(ref mut internal_node, ..) = old_parent_block.node {
+            if old_leaf_index == internal_node.left {
+                internal_node.left = new_internal_node_index;
+            } else if old_leaf_index == internal_node.right {
+                internal_node.right = new_internal_node_index;
+            } else {
+                panic!("child not a child of its parent");
+            }
+        } else {
+            panic!("expected internal node but found leaf");
+        };
+
+        self.insert_entry_to_blob(old_parent_index, &old_parent_block)?;
+
+        self.mark_lineage_as_dirty(old_parent_index)?;
+
+        Ok(new_leaf_index)
+    }
+
+    pub fn batch_insert(
+        &mut self,
+        mut keys_values_hashes: Vec<((KeyId, ValueId), Hash)>,
+    ) -> Result<(), Error> {
+        // OPT: perhaps go back to taking an iterator?
+        // OPT: would it be worthwhile to hold the entire blocks?
+        let mut indexes = vec![];
+
+        if self.block_status_cache.leaf_count() <= 1 {
+            for _ in 0..2 {
+                let Some(((key, value), hash)) = keys_values_hashes.pop() else {
```

Comment:
I take it the insertion order doesn't matter here, because in this special case, you add the last items 2 items first, and then start from the beginning on line 592

Reply:
This is indeed a bit awkward and likely doesn't need to be this way.  But also yes, the order doesn't matter here.  If the caller wants to control the tree structure they need to use another insert mechanism.